packages/go/cypher/model/pgsql/translate/tracking.go

```diff
@@ -0,0 +1,304 @@
+package translate
+
+import (
+	"fmt"
+	"github.com/specterops/bloodhound/cypher/model"
+	"github.com/specterops/bloodhound/cypher/model/pgsql"
+	"strconv"
+)
+
+type IdentifierGenerator map[pgsql.DataType]int
+
+func (s IdentifierGenerator) NewIdentifier(dataType pgsql.DataType) (pgsql.Identifier, error) {
+	var (
+		nextID    = s[dataType]
+		nextIDStr = strconv.Itoa(nextID)
+	)
+
+	// Increment the ID
+	s[dataType] = nextID + 1
+
+	switch dataType {
+	case pgsql.EdgeExpansion:
+		return pgsql.Identifier("ex" + nextIDStr), nil
+	case pgsql.PathComposite:
+		return pgsql.Identifier("p" + nextIDStr), nil
+	case pgsql.NodeComposite:
+		return pgsql.Identifier("n" + nextIDStr), nil
+	case pgsql.EdgeComposite:
+		return pgsql.Identifier("e" + nextIDStr), nil
+	default:
+		return "", fmt.Errorf("identifier with data type %s does not have a prefix case", dataType)
+	}
+}
+
+func NewIdentifierGenerator() IdentifierGenerator {
+	return IdentifierGenerator{}
+}
+
+type Constraint struct {
+	Dependencies pgsql.IdentifierSet
+	Expression   pgsql.Expression
+}
+
+type ConstraintTracker struct {
+	Constraints []*Constraint
+}
+
+func NewConstraintTracker() *ConstraintTracker {
+	return &ConstraintTracker{}
+}
+
+func (s *ConstraintTracker) ConsumeAll() pgsql.Expression {
+	var topLevelExpression pgsql.Expression
+
+	for _, constraint := range s.Constraints {
+		topLevelExpression = OptionalAnd(topLevelExpression, constraint.Expression)
+	}
+
+	s.Constraints = nil
+	return topLevelExpression
+}
+
+func (s *ConstraintTracker) Consume(available pgsql.IdentifierSet) (pgsql.IdentifierSet, pgsql.Expression) {
+	var (
+		matched    = pgsql.IdentifierSet{}
+		constraint pgsql.Expression
+	)
+
+	for idx := 0; idx < len(s.Constraints); {
+		nextConstraint := s.Constraints[idx]
+
+		if available.Satisfies(nextConstraint.Dependencies) {
+			// Remove this constraint
+			s.Constraints = append(s.Constraints[:idx], s.Constraints[idx+1:]...)
+
+			// Append the constraint as a conjoined expression
+			constraint = OptionalAnd(constraint, nextConstraint.Expression)
+
+			// Track which identifiers were satisfied
+			matched.Merge(nextConstraint.Dependencies)
+		} else {
+			// If this constraint isn't satisfied by the available namespace move to the next constraint
+			idx += 1
+		}
+	}
+
+	return matched, constraint
+}
+
+func (s *ConstraintTracker) Constrain(dependencies pgsql.IdentifierSet, constraintExpression pgsql.Expression) {
+	for _, constraint := range s.Constraints {
+		if constraint.Dependencies.Matches(dependencies) {
+			constraint.Expression = OptionalAnd(constraint.Expression, constraintExpression)
+			return
+		}
+	}
+
+	s.Constraints = append(s.Constraints, &Constraint{
+		Dependencies: dependencies,
+		Expression:   constraintExpression,
+	})
+}
+
+type TrackedIdentifier struct {
```

Comment:
i have the same request for this structure as i made for the Constraint structure.  it looks like TrackedIdentifier is a wrapper around a good ol pgsql.Identifier AST.  some comments around the usage of this structure would be grand.  when does an identifier become tracked? what does it mean to be tracked?  what parts of the translation depend on this structure?